code/datums/components/autofire/autofire.dm

```diff
@@ -0,0 +1,137 @@
+/datum/component/automatedfire/autofire
+	///The current fire mode of the shooter
+	var/fire_mode
+	///Delay between two shots when in full auto
+	var/auto_fire_shot_delay
+	///Delay between two burst shots
+	var/burstfire_shot_delay
+	///How many bullets are fired in burst mode
+	var/burst_shots_to_fire
+	///Count the shots fired when bursting
+	var/shots_fired = 0
+	///If the shooter is currently shooting
+	var/shooting = FALSE
+	///If TRUE, the shooter will reset its references at the end of the burst
+	var/have_to_reset_at_burst_end = FALSE
+	///If we are in a burst
+	var/bursting = FALSE
+	///Callback to set bursting mode on the parent
+	var/datum/callback/callback_bursting
+	///Callback to ask the parent to reset its firing vars
+	var/datum/callback/callback_reset_fire
+	///Callback to ask the parent to fire
+	var/datum/callback/callback_fire
+	///Callback to ask the parent to display ammo
+	var/datum/callback/callback_display_ammo
+	///Callback to set parent's fa_firing
+	var/datum/callback/callback_set_firing
+
+/datum/component/automatedfire/autofire/Initialize(_auto_fire_shot_delay = 0.3 SECONDS, _burstfire_shot_delay, _burst_shots_to_fire = 3, _fire_mode = GUN_FIREMODE_SEMIAUTO, datum/callback/_callback_bursting, datum/callback/_callback_reset_fire, datum/callback/_callback_fire, datum/callback/_callback_display_ammo, datum/callback/_callback_set_firing)
```

Comment:
prefer leaving the names without _s and just using src.auto_fire_shot_delay, per https://github.com/cmss13-devs/cmss13/blob/master/.github/guides/STYLES.md#when-passing-vars-through-new-or-initializes-arguments-use-srcvar